### DISCUSSION TITLE
A way to send expo-modules events directly from native main app

### DISCUSSION THREAD

**Discussion by damianfilipek81** [expo/expo#27468] [Archived: Questions]
Hello!
I've got RN app where some part of logic is on native side and some on js part. To communicate between native <=> js we use `emitEvent` method: 

```kt
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter

class ReactNativeEventEmitter(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    init {
        instance = this
    }

    companion object {
        private var instance: ReactNativeEventEmitter? = null

        fun emitEvent(eventName: String, eventData: WritableMap) {
            instance?.reactApplicationContext
                ?.getJSModule(RCTDeviceEventEmitter::class.java)
                ?.emit(eventName, eventData)
                ?: throw IllegalStateException("ReactNativeEventEmitter instance not initialized or reactApplicationContext not available.")
        }
    }

    override fun getName(): String {
        return "EventEmitterModule"
    }
}
```

i want to move to new architecture, but since there is no official documentation on how to send events with turbo modules i would like to use expo modules. Is it possible to create a expo-module that will use [sendEvent](https://docs.expo.dev/modules/module-api/#sendeventeventname-payload) method and then use it in places where we currently use old `RCTDeviceEventEmitter` inside main native project?
Im not sure how exactly expo modules work, do i need to initialize that module somewhere, do i need to create some kind of singleton? I've tried some things but couldn't find any working way.
Also i couldn't find any similar issues or documentation.

Thanks,
Damian

**Comment by tsapeta** [comment]
Hey!

In Expo modules, instead of using `RCTDeviceEventEmitter`, you can use the `EventEmitter` class from `expo-modules-core`. It basically wraps the native module to let you emit and listen to its events.
`expo-battery` might be a good example: https://github.com/expo/expo/blob/main/packages/expo-battery/src/Battery.ts#L11-L13

We're working on improving this even more. In SDK 51, the `EventEmitter` class will be natively-defined in C++ and all modules will already inherit from it out of the box. In other words, this will become possible:

```js
const MyModule = requireNativeModule('MyModule');
MyModule.addListener('my-event', (arg) => { ... });
```

**Reply by damianfilipek81** [reply]
Hi! Thanks for answer,
it looks like you provided me information how to emit and listen events on javascript part. What i'm looking for is to emit events directly from main native project.

**Reply by tsapeta** [reply]
Ah I see, sorry for misunderstanding 😅 Expo modules don't have a dedicated function for this, but you should be able to access `reactApplicationContext` with `appContext.reactContext` in the Expo module. The rest will be the same.

**Reply by damianfilipek81** [reply]
Hi @tsapeta, so i just want confirm that we fully understand each other.
I would like to use expo module's `MyModule` inside my native android project, here's a simple module that emits event `onExpoEvent`.

```kt
package expo.modules.mymodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MyModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("MyModule")

        Events("onExpoEvent")
    }
    fun sendExpoEvent(someData: String) {
        sendEvent("onExpoEvent", mapOf("message" to someData))
    }
}
```

and now, somewhere inside my android logic (my main native project) i would like to use `sendExpoEvent` like: `MyModule().sendExpoEvent("Hello from native")`

now inside a javascript part i would like to listen for that event:
```ts
//MyModule.ts
import { requireNativeModule } from "expo-modules-core";

// It loads the native module object from the JSI or falls back to
// the bridge module (from NativeModulesProxy) if the remote debugger is on.
export default requireNativeModule("MyModule");

//index.ts
import { EventEmitter, Subscription } from "expo-modules-core";
import MyModule from "./src/MyModule";

const emitter = new EventEmitter(MyModule);

function addTestListener(listener: (event) => void): Subscription {
  return emitter.addListener("onExpoEvent", listener);
}
```

when i'm trying to do as above i get error `java.lang.IllegalArgumentException: The module wasn't created! You can't access the app context.`

Now i'm wondering if there's some chance to register(?) somewhere that module? 
So i just want to make sure that it's possible/not possible. 

Thanks, 
Damian

**Reply by damianfilipek81** [reply]
Hey again, @tsapeta any update on that? 
Cheers

**Reply by mokuzuu** [reply]
+1 for this.
I need a way to tell react native part of application when a certain action triggered on native code.

**Reply by Qanpi** [reply]
+1

**Reply by hanclang** [reply]
+1

**Comment by julian-gargicevich** [comment]
+1

How can we trigger an event from the native side with for example something like LiveActivitiesIntents?
